elasticsearch/shard_migration.py

```diff
@@ -0,0 +1,181 @@
+                        }
+                    }
+                ]
+            }
+
+            try:
+                LOG.info(
+                    f"Trying to move shard {shard} of index '{index}' from {source_node} to {target_node}."
+                )
+                headers = {"Content-Type": "application/json"}
+                response = requests.post(
+                    f"{es_host}/_cluster/reroute",
+                    headers=headers,
+                    data=json.dumps(reroute_body),
+                )
+                response.raise_for_status()
+                LOG.info(
+                    f"Shard {shard} of index '{index}' successfully moved from {source_node} to {target_node}."
+                )
+                return True
+            except requests.exceptions.RequestException as e:
+                LOG.warning(
+                    f"Not able to move shard {shard} of index '{index}' to {target_node}. Error: {e}"
+                )
+
+        LOG.warning(
+            f"Shard {shard} of index '{index}' could not be moved to any of the target nodes in the list."
+        )
+
+
+def filter_new_shard_allocation(source_node_ip, es_host):
+    try:
+        LOG.info(f"Filtering new shard allocation to on source node: {source_node_ip}.")
+
+        body = {
+            "persistent": {"cluster.routing.allocation.exclude._ip": source_node_ip}
+        }
+
+        headers = {"Content-Type": "application/json"}
+        response = requests.put(
+            f"{es_host}/_cluster/settings", headers=headers, data=json.dumps(body)
+        )
+        response.raise_for_status()
+        LOG.info(
+            f"Successfully updated cluster settings for filtering new shard allocation on source node {source_node_ip}."
+        )
+    except requests.exceptions.RequestException as e:
+        LOG.error(
+            f"Error updating cluster settings for filtering new shard allocation on source node {source_node_ip}: {e}"
+        )
+
+
+def main():
+    (opts, args) = parse_opts()
+
+    if not opts.source_node or not opts.source_node_ip or not opts.target_nodes:
+        LOG.error(
+            "Options --source-node, --source-node-ip, and --target-nodes are required."
+        )
+        exit(2)
+
+    es_host = opts.es_host
+    source_node = opts.source_node
+    source_node_ip = opts.source_node_ip
+    target_nodes = opts.target_nodes.split(",")
+
+    # Filter new shard allocation on source node
+    filter_new_shard_allocation(source_node_ip, es_host)
+
+    LOG.info(
+        f"Starting shard migration from {source_node} to one of the target nodes: {target_nodes}."
+    )
+
+    # Get indices and shards on the source node
+    indices_on_node = get_indices_on_node(source_node, es_host)
+
+    if indices_on_node:
+        # Gradually move shards to one of the target nodes
+        move_shards(indices_on_node, source_node, target_nodes, es_host)
+    else:
+        LOG.warning(f"No shards found on {source_node}.")
```

Comment:
Kinda messy. Use a class to avoid having to pass `es_node` every time to a function. And move out into functions things that are kinda obvious, but keep the actual flow of the script in `main`:
```suggestion
    es = ElasticSearch(es_host)

    on_source_node = lambda index: index['node'] == source_node

    for index in filter(on_source_node, es.get_indices()):
        es.move_shard(index, source_node, target_nodes)
```
Hiding loops and filtering in functions is adding unnecessary complexity.